app/questionnaire/navigator.py

```diff
@@ -1,193 +1,293 @@
+        :param path: The known path as a list which has been visited already
+        :return: A list of block ids followed through the survey
+        """
+        if block_id in self.CLOSING_INTERSTITIAL_PATH:
+            return path
+
+        this_block = {
+            "block_id": block_id,
+            "group_id": group_id,
+            "group_instance": group_instance,
+        }
+
+        path.append(this_block)
+
+        # Return the index of the block id to be visited
+        block_id_index = next(index for (index, b) in enumerate(blocks) if b["block"]["id"] == block_id and
+                              b["group_id"] == group_id and b['group_instance'] == group_instance)
+
+        block = blocks[block_id_index]["block"]
+
+        if 'routing_rules' in block and len(block['routing_rules']) > 0:
+            for rule in block['routing_rules']:
+                if 'goto' in rule:
+                    should_go = evaluate_goto(rule['goto'], self.answer_store, group_instance)
+                    if should_go is True:
+                        return self.build_path(blocks, group_id, group_instance, rule['goto']['id'], path)
+                    elif should_go is False:
+                        return path
+        elif block_id_index != len(blocks) - 1:
+            next_block_id = blocks[block_id_index + 1]['block']['id']
+            next_group_id = blocks[block_id_index + 1]['group_id']
+            next_group_instance = blocks[block_id_index + 1]['group_instance']
+
+            return self.build_path(blocks, next_group_id, next_group_instance, next_block_id, path)
+        return path
 
-    def get_routing_path(self, answers=None):
+    def get_routing_path(self):
         """
         Returns a list of the block ids visited based on answers provided
         :return: List of block ids
         """
-        answers = answers or {}
         routing_path = []
+        blocks = self.get_blocks()
 
-        return build_path(self.blocks, self.first_block_id, answers, routing_path)
+        return self.build_path(blocks, self.first_group_id, 0, self.first_block_id, routing_path)
 
-    def can_reach_summary(self, answers=None, routing_path=None):
+    def can_reach_summary(self, routing_path=None):
         """
         Determines whether the end of a given routing path can be reached given
         a set of answers
         :param routing_path:
-        :param answers:
         :return:
         """
-        answers = answers or {}
-        routing_path = routing_path or build_path(self.blocks, self.first_block_id, answers, [])
-        last_routing_block_id = routing_path[-1]
+        blocks = self.get_blocks()
+        routing_path = routing_path or self.build_path(blocks, self.first_group_id, 0, self.first_block_id, [])
+        last_routing_block_id = routing_path[-1]['block_id']
 
-        if self.blocks[-1]['id'] == last_routing_block_id:
+        if blocks[-1]['block']['id'] == last_routing_block_id:
```

Comment:
blocks[-1]['block']['id'] might be clearer as summary_block_id = blocks[-1]['block']['id']